File: leetcode/21. Merge Two Sorted Lists/py3/solu.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def mergeTwoLists(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummyhead = curhead = ListNode(-1)
        print('dummy head next: ', dummyhead.next, '\n curhead next, ', curhead.next)
        
        while l1 and l2:
            if l1.val < l2.val:
                curhead.next = l1
                print('curhead.next: ', curhead.next.val)
                l1 = l1.next
            else:
                curhead.next = l2
                print('curhead.next: ', curhead.next.val)
                l2 = l2.next
            curhead = curhead.next
            print('dummy head next: ', dummyhead.next, '\n curhead next, ', curhead.next)
            print('l1: ', l1, '\n l2, ', l2, '\n')


        if l1:
            curhead.next = l1
        if l2:
            curhead.next = l2
        return dummyhead.next

File: leetcode/21. Merge Two Sorted Lists/js/solu.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
 function ListNode(val, next) {
  this.val = (val===undefined ? 0 : val)
  this.next = (next===undefined ? null : next)
}


/**
* @param {ListNode} l1
* @param {ListNode} l2
* @return {ListNode}
*/
var mergeTwoLists = function(l1, l2) {
  let dummyHead = new ListNode(-1, null)
  let curHead = dummyHead
  while (l1 && l2) {
      if (l1.val < l2.val) {
          curHead.next = l1
          l1 = l1.next
      } else {
          curHead.next = l2
          l2 = l2.next
      }
      curHead = curHead.next
      // console.log('l1: ', l1, ', l2: ', l2)
      // console.log('curHead: ', curHead)
      // console.log('dummyHead next: ', dummyHead.next)
  }
  // console.log('dummyHead: ', dummyHead.next)
  if (l1 !== undefined)
      curHead.next = l1
  if (l2) {
      curHead.next = l2
  }
  
  return dummyHead.next
  
};
